TLM/TMPE.API/Manager/ITrafficLightManager.cs

```diff
@@ -2,5 +2,10 @@ namespace TrafficManager.API.Manager {
     using TrafficManager.API.Traffic.Enums;
 
     public interface ITrafficLightManager {
+        bool HasTrafficLight(ushort nodeId);
+
+        bool CanSetTrafficLight(ushort nodeId, bool enabled);
+
+        bool SetTrafficLight(ushort nodeId, bool enabled);
```

Comment:
what is the difference between manual and vanilla (there is no difference)?
why should I control Timed Traffic light in TrafficLightManager?
`public bool SetTrafficLightType(nodeId, TrafficLightType.Timed);` does not make sense at all.
I could actually just  by pass TMPE and instead set/get use `nodeId.ToNode().m_flags.TrafficLight` but that would also bypass TMPE's check for presence of Timed Traffic Light and stuff.

I think it would be better if I changed `CanSetTrafficLight(nodeId, enabled)` and `SetTrafficLight(nodeId, enabled)` to `CanToggleTrafficLight(nodeId)` and `ToggleTrafficLight(nodeId)`. `TrafficLightManager` cannot handle TTL stuff.

Reply:
> what is the difference between manual and vanilla (there is no difference)?

Manual traffic lights are their own thing. I'm honestly unclear on their purpose, but they're definitely not vanilla.

> why should I control Timed Traffic light in TrafficLightManager?

So, I'm not particularly set on doing it the specific way I suggested. The goal I'm trying to achieve is to make any future traffic light API more unified in some way.

Remember that there is no particular reason an API has to follow the implementation structure other than convenience, and sometimes it's better for them to part ways. In this case, the traffic light implementation (particularly timed and manual traffic lights) is dubious at best, and is likely to be rewritten at some point in the future.

But honestly, I think I'm mostly concerned about keeping the traffic light API as small as possible until we have a better sense of what the future will look like. I'd be satisfied with a commitment of that kind, and nothing you've suggested here seems to present any significant problems for future compatibility.

> I think it would be better if I changed `CanSetTrafficLight(nodeId, enabled)` and `SetTrafficLight(nodeId, enabled)` to `CanToggleTrafficLight(nodeId)` and `ToggleTrafficLight(nodeId)`. `TrafficLightManager` cannot handle TTL stuff.

I like your original better. I think that in general, it's best for an API to always do the exact same thing (like enabling or disabling the traffic lights), as opposed to doing something different depending on the current state (like toggling the traffic lights).

But I would suggest dropping the `enabled` argument from `CanSetTrafficLight`. I believe they can always be disabled, so it's just a question of whether they can be enabled.

Reply:
oh you mean the manual traffic light tool. its like timed simulation but manual.
I sometimes use them to clear traffic. It feels like acting like traffic police sent to help with traffic :)

>  I believe they can always be disabled, so it's just a question of whether they can be enabled.

They can't be disabled if they are TTL ( i mean they can but to prevent accidentally disabling them it is done trough dedicated API that also sends a warning to user).

Lets say there is TTL at a node and someone calls CanSetTrafficLight(nodeId, true). I am not sure if I can return true because the result is not a normal traffic light but a timed traffic light. if I ask `CanToggleTrafficLight(nodeId)` on the other hand I remove that ambiguity. 

What do you think?

Reply:
All these details and how messy they get is why I said:

> But honestly, I think I'm mostly concerned about keeping the traffic light API as small as possible until we have a better sense of what the future will look like. I'd be satisfied with a commitment of that kind, and nothing you've suggested here seems to present any significant problems for future compatibility.

I'm now thinking that the best short-term plan for traffic lights is extreme minimalism and avoiding things so implementation-specific that they inhibit future development, all while assuming that anything created today will probably get deprecated. I think your original proposal on the traffic light stuff meets that standard.

The thing is that all the stuff you're describing is catering to the way traffic lights are currently implemented. It has nothing to do with how real traffic lights work. Real traffic lights start with sets of traffic movements organized according to which can occur together without conflict. From there, a number of schemes can be followed based on timers, sensors, or a combination. A real traffic light can be switched between these at will as long as all the right equipment is in place. It often even varies depending on time of day. (Maybe a use case for Real Time integration?)

Any major enhancements, if I had my way, would be along these lines, and there's just no way to plan for that with what we have today. So I think I'm just going to "resolve" this part. 🙂